examples/QIR/Simulation/README.md

```diff
@@ -0,0 +1,193 @@
+
+---
+
+`IRuntimeDriver`
+
+```cpp
+// Qubit management
+virtual Qubit AllocateQubit() = 0;
+virtual void ReleaseQubit(Qubit qubit) = 0;
+virtual std::string QubitToString(Qubit qubit) = 0;
+
+// Result management
+virtual void ReleaseResult(Result result) = 0;
+virtual bool AreEqualResults(Result r1, Result r2) = 0;
+virtual ResultValue GetResultValue(Result result) = 0;
+virtual Result UseZero() = 0;
+virtual Result UseOne() = 0;
+```
+
+---
+
+`IQuantumGateSet`
+
+```cpp
+// Elementary operations
+virtual void X(Qubit target) = 0;
+virtual void Y(Qubit target) = 0;
+virtual void Z(Qubit target) = 0;
+virtual void H(Qubit target) = 0;
+virtual void S(Qubit target) = 0;
+virtual void T(Qubit target) = 0;
+virtual void R(PauliId axis, Qubit target, double theta) = 0;
+virtual void Exp(long numTargets, PauliId paulis[], Qubit targets[], double theta) = 0;
+
+// Multi-controlled operations
+virtual void ControlledX(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledY(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledZ(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledH(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledS(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledT(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledR(long numControls, Qubit controls[], PauliId axis, Qubit target, double theta) = 0;
+virtual void ControlledExp(long numControls, Qubit controls[], long numTargets, PauliId paulis[], Qubit targets[], double theta) = 0;
+
+// Adjoint operations
+virtual void AdjointS(Qubit target) = 0;
+virtual void AdjointT(Qubit target) = 0;
+virtual void ControlledAdjointS(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledAdjointT(long numControls, Qubit controls[], Qubit target) = 0;
+
+// Measurement operations
+virtual Result Measure(long numBases, PauliId bases[], long numTargets, Qubit targets[]) = 0;
+```
+
+---
+
+`IDiagnostics`
+
+```cpp
+// Internal state
+virtual void GetState(TGetStateCallback callback) = 0; // deprecated
+virtual void DumpMachine(const void* location) = 0;
+virtual void DumpRegister(const void* location, const QirArray* qubits) = 0;
+
+// Measurement outcomes
+virtual bool Assert(long numTargets, PauliId bases[], Qubit targets[], Result result, const char* failureMessage) = 0;
+virtual bool AssertProbability(long numTargets, PauliId bases[], Qubit targets[], double probabilityOfZero, double precision, const char* failureMessage) = 0;
+```
+
+---
+
+The requirements on a simulator are then:
+
+- Qubit management, which can include:
+  - virtual qubits
+  - logical qubits
+  - physical qubits
+
+  Generally, the physical qubit layer is not present in simulators, as they typically directly implement logical qubits, unless the simulator is built to simulate noise in real quantum computers.
```

Comment:
It may be worth mentioning the noisy simulators as a counterexample.

Reply:
Are you referring to one of our existing simulators?